### PR TITLE
Final commit LockThis :analyzer + code fix provider + tests

### PR DESCRIPTION
**Pull request by Kavignon**
Separated the analyzer in two different code action. It looks and feels more logical to process this way.

TryGetDiagnosticFromLockStatement focuses on lock statement and verifies if the statement inside the lock is of the kind of ThisExpression

TryGetDiagnosticFromPossibleSynchronizeAttribute focuses on attribute syntax. It goes through the attribute syntax to find out if the Synchronized enumeration value was accessed.

The code fix provider was updated.

### CODE REVIEW

File: NR6Pack/Diagnostics/Custom/LockThisCodeFixProvider.cs [hunk 1]
```diff
@@ -24,41 +24,45 @@
...
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var span = context.Span;
+            var diagnostics = context.Diagnostics;
+            var root = await document.GetSyntaxRootAsync(cancellationToken);
+            var diagnostic = diagnostics.First();
+            var node = root.FindNode(context.Span);
+            if (node == null)
+                return;
+            var newRoot = root.RemoveNode(node, SyntaxRemoveOptions.KeepNoTrivia);
```
Comment (mkrueger) [inline NR6Pack/Diagnostics/Custom/LockThisCodeFixProvider.cs]: This just removes the whole node - looking at the old logic this is not correct.

**Comment by Kavignon** [comment]
I have been looking for a way to do it but I don't understand how to do it. I don't really get the logic after i have created the new node.

**Comment by mkrueger** [comment]
Basically it's just an AST replacement, Do it step by step - just replace the 'this' expression first.

The old logic is an editor operation and the new one is an AST transformation. However the old editor operations used an API like an AST transformation therefore it should be straightforward.

You can look at the other CodeFixProviders - there are plenty of examples.

**Comment by Kavignon** [comment]
I'll do my best to have it done by tomorrow. 
One thing, how can I test the provider to make sure it works ?

**Comment by mkrueger** [comment]
With the unit tests.​

The Analyze method can test the output of the code fix provider as well.
It's an all in one solution :)

**Comment by Kavignon** [comment]
I've been at it for the past week but i really dont get how to write a code fix

**Comment by mkrueger** [comment]
Just try a simple AST transformation and go with that.

Look at that: 

https://github.com/icsharpcode/NRefactory/blob/roslyn/NR6Pack/Diagnostics/Synced/PracticesAndImprovements/ConvertIfDoToWhileCodeFixProvider.cs

First replace this with a simple name to the newly generated variable and then just add a field to the type.

For inserting a field just search for a refactoring that does it - for example:

https://github.com/icsharpcode/NRefactory/blob/roslyn/NR6Pack/CodeRefactorings/Synced/ReplaceAutoPropertyWithPropertyAndBackingFieldCodeRefactoringProvider.cs

**Comment by Kavignon** [comment]
thanks mike.
ill sort this out. 
make the code fix for rewrite + new tests
separate everything in two branches and try to make 2 p.r by sunday

**Comment by Kavignon** [comment]
The best I came up with, end that is not even compiling is :

**Comment by Kavignon** [comment]
context.RegisterCodeFix(CodeActionFactory.Create(node.Span, diagnostic.Severity, "Use locking statements with an object for locking", token =>
            {
                var newNode = SyntaxFactory.LockStatement()
                var newRoot = root.ReplaceNode((SyntaxNode)node, newNode.(node.GetLeadingTrivia()).WithAdditionalAnnotations(Formatter.Annotation));
                return Task.FromResult(document.WithSyntaxRoot(newRoot));
            }), diagnostic);

**Comment by Kavignon** [comment]
I have tried to make sense of the code fix providers, but I don't seem to be able to get my head around with it at the moment. I'm going to continue working on it

**Pull request closed without merging**